### DISCUSSION TITLE
EasyBuild for Tinker9

### DISCUSSION THREAD

**Discussion by jacekkozuch** [TinkerTools/tinker-gpu#171] [General]
Hi,

I've seen that Tinker8 can be build using EasyBuild. Is it planned from your side make an EasyBuild config available for Tinker9? 

If yes, this would help a lot to get it build on the HPC at the institute where I'm working and make it available to every user.

Thanks and Best,
Jacek

**Comment by zhi-wang** [comment]
Let me do some research on EasyBuild. I am not familiar with this tool. I'll get back to you in a while.

**Comment by jacekkozuch** [comment]
Hi,

Just wanted to ask if you found out something about EasyBuild? 
Thx!

Jacek

**Reply by zhi-wang** [reply]
Sorry for the late reply. We (mostly I) think that the current building procedure with CMake is already very complex due to the mixed usage of 3 different compilers. Supporting another build system is probably not a priority at this point.

**Comment by tardigradus** [comment]
[EasyBuild](https://easybuild.io/) is not another build system.  It is more of a meta build system which knows allow you to write a kind of recipe (a so-called easyconfig file) which enables you to build a piece of software and any dependencies it may have.  It relies on so-called easyblocks which are modules with encapsulate ways of installing software.  So there are easyblocks for CMake, configure/make, binaries, etc, but also some which are specific to a particular piece of software if this has an unusual and/or complex installation mechanism.  The easyconfigs and easyblocks are stored on Github, so once the work has been done, anyone can use it. 

EasyBuild thus makes it very easy to install a large amount of software very quickly and without you needing to know any details of the installation process.  This is particularly useful for small HPC sites which may not have the resources to install complex software for users unless it is available via EasyBuild.  Easyconfigs exist for version 8 of Tinker, so there is potentially already interest and know-how available for Tinker9 within the EasyBuild community.

**Comment by zhi-wang** [comment]
Thank you very much for your input. What are your thoughts on the Docker images? If we (TinkerTools) can set up a continuously integrated service to automatically build and release the executables in the form of Docker images, do you think this will suffice your research and make using Tinker9 even easier? Thanks!

**Reply by tardigradus** [reply]
Docker images would also be useful, although on HPC systems Docker is often not used.  Instead, similar container systems, such as Apptainer/Singluarity, are used, which can convert Docker images to their own format.  However, one of the features of EasyBuild is that, because you are building from source, you can build software optimized for specific CPUs. You would loose this performance advantage with a container.

Depending on how different the build process for Tinker9 is from that of previous versions, it might not be too much work to get it working in EasyBuild.  The [EasyConfig](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/t/TINKER/TINKER-8.8.1-foss-2020a.eb) just specifies things like the version, where the software is located and checksums for the source packages.  The corresponding [EasyBlock](https://github.com/easybuilders/easybuild-easyblocks/blob/develop/easybuild/easyblocks/t/tinker.py) encapsulates the actual steps of the build process. One of the people involved with the previous EasyConfigs might be interested in creating one for Tinker9.
